pipeline/stages/test.yaml

```diff
@@ -0,0 +1,157 @@
+                rgName="${service_name}rg"
+
+                allOk=true
+                for i in {1..2}
+                do
+
+                    vmName="${service_name}vm${i}"
+                    echo "Testing $vmName in $rgName."
+
+                    #az vm show --resource-group $rgName --name $vmName
+                    passwordDisabled=$(az vm show --resource-group $rgName --name $vmName --query "osProfile.linuxConfiguration.disablePasswordAuthentication" -o tsv)
+                    if [ "$passwordDisabled" = "false" ]
+                    then
+                        echo "##vso[task.LogIssue type=error;]$vmName in $rgName has password access enabled."
+                        allOk=false
+                    fi
+                    secretCount=$(az vm show --resource-group $rgName --name $vmName --query "length(osProfile.secrets[0].vaultCertificates)" -o tsv)
+                    if [ "$secretCount" = "0" ]
+                    then
+                        echo "##vso[task.LogIssue type=error;]There are no Key Vault Secrets attached to $vmName in $rgName"
+                        allOk=false
+                    else
+                        certCount=$(az vm show --resource-group $rgName --name $vmName --query "length(osProfile.secrets[0].vaultCertificates)" -o tsv)
+                        if [ "$certCount" = "0" ]
+                        then
+                            echo "##vso[task.LogIssue type=error;]There are no Key Vault Secret Certificates attached to $vmName in $rgName"
+                            allOk=false
+                        fi
+                    fi
+
+                done
+
+                if [ "$allOk" = "false" ]
+                then
+                    echo "##vso[task.LogIssue type=error;]There are issues with vms in $rgName"
+                    exit 1
+                fi
+          - task: AzureCLI@2
+            name: 'certificate'
+            displayName: Check Machine Certificates
+            inputs:
+              azureSubscription: '${{variables.subscriptionName}}'
+              scriptType: 'bash'
+              scriptLocation: 'inlineScript'
+              inlineScript: |
+                service_name="${{variables.product}}-recordings-${{parameters.env}}-"
+                rgName="${service_name}rg"
+                nsgName="${service_name}sg"
+                ruleName="pipelineTestRule"
+
+                echo "Get Source IP"
+                sourceIp="$(dig @resolver4.opendns.com myip.opendns.com +short)"
+                echo "Source IP: $sourceIp"
+
+                echo "Create NSG rule"
+                az network nsg rule create --name $ruleName --nsg-name $nsgName --priority '200' --resource-group $rgName --access 'Allow' --description "Test from the Deployment Pipeline"  --destination-port-ranges '443' --direction 'Inbound' --protocol 'Tcp' --source-address-prefixes $sourceIp  --source-port-ranges '*'
+
+                sleep 20s
```

Comment:
and conversly, I've had issues with security group rules taking up to a minute to apply. Reckon this needs a check added here? Or maybe just a comment saying rerun if this fails? If you check some the azure-devops-templates/jobs/azWebAppdeloyJob the firewall adding/removal there is a bit OTT

Reply:
Loop add round check for nsg rule to be added